[IMP] website: facilitate searching for industries in the configurator

bso-odoo opened the pull request:
Before this commit the industries were only matched on starting with or
containing the searched term and the autocompletion results were shown
alphabetically.

After this commit the industries are also matched if they contain all
individual search terms.
All terms are matched against the label and each synonym (if any exists),
but the industry is only displayed once, with its label.
If there are 7 results or less, the results are alphabetically sorted on
the label, otherwise the IAP result order is kept (which is expected to
be by descending cross-theme total hit count).
The truncation of the matching list to 15 results gives higher priority
to labels with the least words (so that "restaurant" still appears even
if there are 15 more common sub-types).

task-2688878

Related to https://github.com/odoo/iap-apps/pull/443

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/82018).

bso-odoo commented:
@rdeodoo is welcome to review this PR.
Note that it is related to https://github.com/odoo/iap-apps/pull/443.

rdeodoo reviewed (changes requested):
About the new synonyms field, why not just comma?
> Distinct synonyms are expected to be separated by commas, pipes or a line feed.

bso-odoo commented:
> About the new synonyms field, why not just comma?
> 
> > Distinct synonyms are expected to be separated by commas, pipes or a line feed.

Synonyms are likely to be imported from CSV, where list values within a single column are often separated by pipes - this was an agreement with PDE who will actually handle the synonyms.
I also included "new line" in case this is used in the text area when modifying the synonyms manually from the UI.

bso-odoo commented on addons/website/static/src/components/configurator/configurator.js:
```diff
@@ -179,23 +179,39 @@ class DescriptionScreen extends Component {
+        if (matches.length > limit) {
+            // Keep matches with the least number of words so that e.g. "restaurant" remains
+            // available even if there are 15 specific sub-types that have a higher hit count.
+            const matchesPerWordCount = [...matches].sort((x, y) => x.wordCount - y.wordCount);
+            const survivors = matchesPerWordCount.slice(0, limit).map((val) => val.label);
+            results = matches.filter((val) => survivors.includes(val.label));
```
It is about keeping survivors in their original order, therefore we need to first establish the list of survivors (1 & 2), then we can filter the correctly sorted values (3).
This is because the priority rule for being a survivor is itself about a different sorting of the values.

rdeodoo replied:
I was conceptually more concern about the fact that we could filter out too many results or even all the results by doing so? 

Either the (3) (filter out some entries) is useless (as always true and not filtering anything) or it can filter out all the results?

I might be missing something

rdeodoo reviewed (approved):
This was IMHO over complicated.

I rewrote both part (the first loop and the if condition).
There is no more possible misunderstanding in the `if` condition as it is now a one liner thanks to the use of a new property `hitCountOrder`, which was not sent by IAP and was forcing you to reorder through a filter (that's where I was lost).

- I removed the nested `for` loop with a custom exit `continue synonymLoop;
- Reversed the logic "synonym > term" to "term > synonym" to have a direct return statement instead of "dodging" it by continuing the loop
- no more need of `results` variable, `matchesPerWordCount` variable, `survivors` variable
- added come comment that really help figuring stuff

Let me know if I broke something or if you want to apply another change

rdeodoo commented on addons/website/static/src/components/configurator/configurator.js:
```diff
@@ -184,25 +184,42 @@ class DescriptionScreen extends Component {
+                    return false;
+                }
+            }
+            val.wordCount = val.label.split(" ").length;
+            val.hitCountOrder = index;
```
Note that those lines should probably be moved inside the `getInitialState` to alter it after the RPC only once, and not on every autocomplete.

bso-odoo commented:
> Reversed the logic "synonym > term" to "term > synonym" to have a direct return statement instead of "dodging" it by continuing the loop

That will return different matches: instead of keeping results where at least one synonym matches all terms, it will keep results where all terms appear across every synonyms. If we want to do this, we could also just have one single string with everything inside.

rdeodoo commented:
Well then we could just replace
```js
for (const term of terms) {
    if (!candidates.some(el => el.includes(term))) {
        return false;
    }
}
```
by
```js
for (const candidate of [val.label, ...(val.synonyms || '').split(" ")]) {
    if (terms.every(term => candidate.includes(term))) {
        return true;
    }
}
```
to restore the initial behavior I missed but keep the same change?

I also applied the improvement to add the object properties in the init

[diff](https://github.com/odoo/odoo/compare/57ae5e426e76af118fb94690dcc164768dd61c46..7a73c88f401904e833011e858615e172de2ae115)

bso-odoo commented:
> ```js
> for (const candidate of [val.label, ...(val.synonyms || '').split(" ")]) {
> ```

If the synonyms are split on space, that will isolate each of their words - while keeping together the last and first words of successive synonyms.
The initial intent was to either allow the import of synonyms from CSVs, using a pipe character between values (`|`), or to allow an interactive user on the IAP instance to fill in the text area by putting one synonym per line, or separate them with commas.
=> `.split(/[|,\n]+/)])`

rdeodoo commented:
That `split(" ")` shouldn't have been part of my refactoring, not sure how it went through :wink:

bso-odoo commented:
Updated the method documentation which was still describing the old implementation's results.

rdeodoo reviewed (approved):
Waiting for the IAP PR to be merged then we ship this one.
